File: cli/src/commands/slice.js
import path from 'path';
import { loadGraph } from '../graph.js';
import { generateOverview, generateSlices, getModuleSliceWithDeps } from '../slicer.js';

/**
 * Register the `slice` command on the given Commander program.
 *
 * @param {import('commander').Command} program - The Commander program instance.
 */
export function registerSliceCommand(program) {
  program
    .command('slice [module]')
    .description('Output module slice or overview as JSON')
    .option('--with-deps', 'Include dependency info in module slice')
    .option('--dir <dir>', 'Project directory', '.')
    .action(async (moduleName, options) => {
      const rootDir = path.resolve(options.dir);
      const outputDir = path.join(rootDir, '.codemap');

      let graph;
      try {
        graph = await loadGraph(outputDir);
      } catch {
        console.error('No code graph found. Run "codegraph scan" first.');
        process.exit(1);
      }

      if (!moduleName) {
        // No module specified: output overview JSON
        const overview = generateOverview(graph);
        console.log(JSON.stringify(overview, null, 2));
        return;
      }

      // Module specified
      if (!graph.modules[moduleName]) {
        console.error(`Module "${moduleName}" not found in graph.`);
        console.error(`Available modules: ${Object.keys(graph.modules).join(', ')}`);
        process.exit(1);
      }

      if (options.withDeps) {
        // Include dependency info
        const sliceWithDeps = getModuleSliceWithDeps(graph, moduleName);
        console.log(JSON.stringify(sliceWithDeps, null, 2));
      } else {
        // Simple module slice
        const slices = generateSlices(graph);
        const slice = slices[moduleName];
        console.log(JSON.stringify(slice, null, 2));
      }
    });
}

File: rust-cli/src/commands/slice.rs
use clap::Args;
use std::path::PathBuf;

#[derive(Args)]
pub struct SliceArgs {
    /// Module name (omit for overview)
    pub module: Option<String>,
    /// Include dependency info in module slice
    #[arg(long)]
    pub with_deps: bool,
    /// Project directory
    #[arg(long, default_value = ".")]
    pub dir: String,
}

pub fn run(args: SliceArgs) {
    let root = PathBuf::from(&args.dir);
    let root = match root.canonicalize() {
        Ok(p) => p,
        Err(e) => {
            eprintln!("Error: cannot resolve directory '{}': {}", args.dir, e);
            std::process::exit(1);
        }
    };

    let codemap_dir = root.join(".codemap");
    let graph = match crate::graph::load_graph(&codemap_dir) {
        Ok(g) => g,
        Err(e) => {
            eprintln!(
                "Error: could not load graph from {}/.codemap/: {}",
                root.display(),
                e
            );
            eprintln!("Run 'codegraph scan {}' first.", root.display());
            std::process::exit(1);
        }
    };

    match args.module {
        None => {
            // 输出 overview
            let overview = crate::slicer::generate_overview(&graph);
            match serde_json::to_string_pretty(&overview) {
                Ok(json) => println!("{}", json),
                Err(e) => {
                    eprintln!("Serialization error: {}", e);
                    std::process::exit(1);
                }
            }
        }
        Some(mod_name) => {
            if args.with_deps {
                match crate::slicer::get_module_slice_with_deps(&graph, &mod_name) {
                    Ok(slice) => match serde_json::to_string_pretty(&slice) {
                        Ok(json) => println!("{}", json),
                        Err(e) => {
                            eprintln!("Serialization error: {}", e);
                            std::process::exit(1);
                        }
                    },
                    Err(e) => {
                        eprintln!("Error: {}", e);
                        std::process::exit(1);
                    }
                }
            } else {
                match graph.modules.get(&mod_name) {
                    Some(mod_data) => {
                        let slice =
                            crate::slicer::build_module_slice(&graph, &mod_name, mod_data);
                        match serde_json::to_string_pretty(&slice) {
                            Ok(json) => println!("{}", json),
                            Err(e) => {
                                eprintln!("Serialization error: {}", e);
                                std::process::exit(1);
                            }
                        }
                    }
                    None => {
                        eprintln!("Error: module \"{}\" not found in graph.", mod_name);
                        std::process::exit(1);
                    }
                }
            }
        }
    }
}
